### PR TITLE
CATROID-1103 Refactor ProjectSaveTask to kotlin

### PR DESCRIPTION
**Pull request by blindguardian50**
https://jira.catrob.at/browse/CATROID-1103

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Review by petbuer — APPROVED** [review]

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaver.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            }
+        }
+    }
+
+    interface ProjectSaveListener {
+        fun onSaveProjectComplete(success: Boolean)
+    }
+}
+
+fun saveLegoSettings(project: Project?, context: Context): Boolean {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaver.kt:63]: Please rename to saveProjectSerial (or smth similar)

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaver.kt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    project.settings.filterIsInstance<LegoNXTSetting>().forEach { setting ->
+        setting.updateMapping(sensorMapping)
+        return
+    }
+    project.settings.add(LegoNXTSetting(sensorMapping))
+}
+
+private fun saveLegoEV3SettingsToProject(project: Project, context: Context) {
+    if (!project.requiredResources.contains(Brick.BLUETOOTH_LEGO_EV3)) {
+        project.settings.toTypedArray().filterIsInstance<LegoEV3Setting>().forEach { setting ->
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaver.kt:88]: Nice 👍

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaver.kt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+class ProjectSaver(private val project: Project, context: Context) {
+    private val weakContextReference: WeakReference<Context> = WeakReference(context)
+    private var onProjectSaveListener: ProjectSaveListener? = null
+    private val scopeIO = CoroutineScope(Dispatchers.IO)
+
+    fun setOnProjectSaveListener(listener: ProjectSaveListener?) {
+        onProjectSaveListener = listener
+    }
+
+    fun saveProject() {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaver.kt:47]:
```suggestion
    fun saveProjectAsync(listener: ProjectSaveListener?) {
```
Please rename. Dependency injection of the listener is not necessary, but please do it if possible. (It also reduces the probability of bugs and memory leaks (If an activity is passed or some other context)
If you want you can get rid of the ProjectSaveListener with "onSaveProjectComplete: (Boolean) -> Unit"

It's also possible to dependency inject the scope:
```suggestion
    fun saveProject(scope: CoroutineScope, onSaveProjectComplete: (Boolean)->Unit) {
```
in the tests, you can then use runBlocking.

**Review by gPathpp — CHANGES_REQUESTED** [review]

**Review by gPathpp — APPROVED** [review]
LGTM! 👍

**Pull request merged**